### DISCUSSION TITLE
Malformed Log File

### DISCUSSION THREAD

**Discussion by ladywhiskers** [beetbox/beets#4937] [General]
I'm doing my best to work out beets as a complete novice, and I've had a lot of luck so far! However, I am now trying to import skipped directories from my logfile, and I'm not sure where to go from here.

I'm getting the error  `error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths` which, after a bit of research, I think means that the logfile contains a mix of directories (like `F:\Music\blah blah` and `\Music\blah blah`. I've had a look through the log file and I can't see this occurring anywhere in the file though.

Would appreciate being pointed in the right direction! I've attached the logfile (saved it as .txt first) and below is a copy of what Powershell is spitting out at me:

```
PS C:\Users\laura> beet import --from-logfile=LOGFILE
discogs: python3-discogs-client version should be >= 2.3.15
error: unreadable logfile LOGFILE: [Errno 2] No such file or directory: '\\\\?\\C:\\Users\\laura\\LOGFILE'
PS C:\Users\laura> beet import --from-logfile=F:/Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura> beet import --from-logfile F:/Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura>  beet import --from-logfile /Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: unreadable logfile /Music/BEETS_IMPORTER_LOG: [Errno 2] No such file or directory: '\\\\?\\C:\\Music\\BEETS_IMPORTER_LOG'
PS C:\Users\laura> beet import --from-logfile F:/Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura> beet import --from-logfile F:Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura> beet import --from-logfile "F:/Music/BEETS_IMPORTER_LOG"
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura> beet import --from-logfile /Music/BEETS_IMPORTER_LOG
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile /Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
PS C:\Users\laura>  beet import --from-logfile "F:/Music/BEETS_IMPORTER_LOG"
discogs: python3-discogs-client version should be >= 2.3.15
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
```
[BEETS_IMPORTER_LOG.txt](https://github.com/beetbox/beets/files/12836231/BEETS_IMPORTER_LOG.txt)

**Comment by sampsyo** [comment]
Thanks for reporting this! Looking into the error handling for this log-file parser, it looks like we are doing a pretty crude job with catching and reporting errors:
https://github.com/beetbox/beets/blob/e10b955a931e4c205b0cadf0860797c0aeee736c/beets/ui/commands.py#L107-L111

A `ValueError` can arise there for many different reasons, and we are just propagating the error message without more details about where it came from.

It should be possible to learn more from the traceback in verbose mode. Trying `beet -vv import […]` should give a little more detail. However, I am virtually certain that the error is coming from the `os.commonpath` call on this line:
https://github.com/beetbox/beets/blob/e10b955a931e4c205b0cadf0860797c0aeee736c/beets/ui/commands.py#L99

And in turn, the problem with _that_ is that it's just splitting the end of the log-file line on `;` characters and hoping that this yields a valid list of paths. Pulling from your log file, that makes sense for lines like this:

```
skip F:\Music\The xx\xx; F:\Music\The xx\xx\CD 01; F:\Music\The xx\xx\CD 02
```

…because we actually did log three different paths, separated by `;` characters. But you have other lines in your log where there is just one path that happens to contain a `;` character, like this one:

```
skip F:\Music\Beethoven; Sydney Symphony\Historic Sydney Symphony Performances
```

In conclusion, this is definitely a bug in beets. The log file parser can't distinguish between `;`s that occur inside paths and ones that separate paths. I'm not sure what to do about this (probably the log-file emitter has to change somehow), but would you perhaps be willing to file a bug to indicate that we have a problem to address here?

To work around this for now, you might consider manually editing your log file and deleting lines where `;` occurs in a path.

**Reply by ladywhiskers** [reply]
I am genuinely surprised that this wasn't an issue of me mistyping something in the commands! I'll file a bug for sure, thank you for taking the time to take me through what's going on here.

Running in verbose, I got the following:
```
PS C:\Users\laura> beet -vv import --from-logfile "F:/Music/BEETS_IMPORTER_LOG"
user configuration: C:\Users\laura\AppData\Roaming\beets\config.yaml
data directory: C:\Users\laura\AppData\Roaming\beets
plugin paths: C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beets-1.6.1-py3.10.egg\beetsplug; C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beetsplug; C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beets-1.6.1-py3.10.egg\beetsplug; C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beetsplug\yearfixer; C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beetsplug\autofix; C:\Users\laura\AppData\Local\Programs\Python\Python310\Lib\site-packages\beetsplug\bandcamp
lastgenre: Loading canonicalization tree c:\users\laura\appdata\local\programs\python\python310\lib\site-packages\beets-1.6.1-py3.10.egg\beetsplug\lastgenre\genres-tree.yaml
discogs: python3-discogs-client version should be >= 2.3.15
fetchart: google: Disabling art source due to missing key
fetchart: lastfm: Disabling art source due to missing key
Sending event: pluginload
library database: F:\Music\musiclibrary.db
library directory: F:\Music
Sending event: library_opened
error: malformed logfile F:/Music/BEETS_IMPORTER_LOG: Can't mix absolute and relative paths
```

Interestingly, doing what you suggested resulted in beets giving me a much more specific (and easy to fix) error `error: malformed logfile F:/Music/BEETS_IMPORTER_LOG2: line 1441 is invalid` - I had a blank line in the file. Once that was fixed, the logfile has now imported with no issue!
